modules/seqexec/server/src/main/scala/seqexec/server/ghost/GHOST.scala

```diff
@@ -65,51 +67,53 @@ object GHOST {
 
   val sfName: String = "GHOST"
 
-  // We always want a GHOSTConfig at this point, so don't use a for comprehension as not all parameters will be
-  // present and we don't want the for to bork prematurely with a missing key error.
   def fromSequenceConfig[F[_]: Sync](config: Config): SeqActionF[F, GHOSTConfig] = {
     def extractor[A : ClassTag](propName: String): Option[A] =
       config.extractAs[A](INSTRUMENT_KEY / propName).toOption
 
+    def angleExtractor[A <: Angle](fmt: Format[String, A])(propName: String): Either[ExtractFailure, Option[A]] = {
+      // 1. content = None: nothing to process, so Right(None).
+      // 2. process content = Some(a) indicating success, so Right(Some(a))
+      // 3. process content = None indicating failure, so Left(error)
+      val content: Option[String] = extractor[String](propName)
+      val result: Option[Option[A]] = content.map(fmt.getOption)
+      result.map {
+        case None       => Left(ConversionError(INSTRUMENT_KEY / propName,
+                                                s"Could not parse $propName content: ${content.getOrElse("")}"))
+        case other      => Right(other)
+      }.getOrElse(Right(None))
```

Comment:
I'm sure there's a much nicer way to do this that I overlooked, so by all means, please teach me what it is :-).

Reply:
Is there a `toRight(leftValue)` in cats?

Reply:
I think you could say

```scala
def formatExtractor[A](
  fmt: Format[String, A]
): String => Either[ExtractFailure, Option[A]] = { propName =>
  extractor[String](propName) match {
    case None    => Right(None)
    case Some(c) => fmt.getOption(c).toRight(ConversionError(...))
  }
}
```
The `Angle` upper bound seems unnecessary, and if you return a function type you won't need the η-expansion operator `_` on `raExtractor` and `decExtractor` below. You could also write it as `.fold` instead of `match` but I think `match` is easier to read here.

Reply:
I have not tested this, it may not even typecheck.

Reply:
Thanks, @tpolecat. I'll give that a try after the seqexec meeting is done.